scanner/unknown.go

```diff
@@ -0,0 +1,209 @@
+package scanner
+
+import (
+	"io/fs"
+	"io/ioutil"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/bitrise-io/go-utils/sliceutil"
+
+	"github.com/bitrise-io/go-utils/log"
+)
+
+const maxDepth = 3
+
+type UnknownToolDetector interface {
+	// ToolName is the human-readable name of a tool/framework/platform
+	ToolName() string
+
+	// DetectToolIn should search recursively in rootPath looking for the given tool
+	DetectToolIn(rootPath string) (DetectionResult, error)
+}
+
+type DetectionResult struct {
+	Detected    bool
+	ProjectTree string
+}
+
+var UnknownToolDetectors = []UnknownToolDetector{
+	singleFileToolDetector{toolName: "Tuist", file: "Project.swift"},
+	singleFileToolDetector{toolName: "Xcodegen", file: "project.yml"},
+	singleFileToolDetector{toolName: "Swift Package Manager", file: "Package.swift"},
+	multiFileToolDetector{toolName: "Buck", primaryFile: "BUCK", optionalFiles: []string{".buckversion", ".buckconfig", ".buckjavaargs"}},
+	kotlinMultiplatformDetector{},
+}
+
+var excludedDirs = []string{
+	".git",
+	".idea",
+	"node_modules",
+	"Pods",
+	"Carthage",
+	"CordovaLib",
+	"Framework",
+}
+
+type singleFileToolDetector struct {
```

Comment:
The `singleFileToolDetector` and `multiFileToolDetector` implementations are too similar to each other which made me think we could unify them. Like a `singleFileToolDetector` can be represented with a `multiFileToolDetector` with a single required item and an empty optional items list. 

And if we do that then we can drop the first part of its name and just keep `toolDetector`.

Reply:
That's a good idea actually!